jsonschema/util.go

```diff
@@ -455,6 +455,46 @@ func fieldJSONInfo(f reflect.StructField) jsonInfo {
 	return info
 }
 
+// walkStructProperties traverses a struct's fields using the same precedence rules as encoding/json.
+// It calls the provided visit function for each discovered field. The StructField passed to the
+// visit function will have its Index field populated with the full path from the top-level struct.
+func walkStructProperties(t reflect.Type, visit func(sf reflect.StructField) error) error {
+	var walk func(reflect.Type, []int) error
+	walk = func(typ reflect.Type, path []int) error {
+		// First Pass: Process fields that are not anonymous pointers to structs.
+		for i := range typ.NumField() {
+			sf := typ.Field(i)
+			fieldType := sf.Type
+
+			if sf.Anonymous && sf.IsExported() && fieldType.Kind() == reflect.Pointer && fieldType.Elem().Kind() == reflect.Struct {
+				continue
+			}
+
+			// Create a copy of the field and set its full index path.
+			fieldCopy := sf
+			fieldCopy.Index = append(slices.Clone(path), i)
+
+			if err := visit(fieldCopy); err != nil {
+				return err
+			}
+		}
+
+		// Second Pass: Recurse into and flatten anonymous pointer-to-struct fields.
+		for i := range typ.NumField() {
```

Comment:
Collect embedded (anonymous) fields in the first pass, to avoid the double loop.